XLS-0056d-batch/README.md

```diff
@@ -0,0 +1,503 @@
+
+Each object in this array contains the following fields:
+
+|FieldName | Required? | JSON Type | Internal Type |
+|:---------|:-----------|:---------------|:------------|
+|`Account`|✔️|`string`|`STAccount`|
+|`SigningPubKey`| |`string`|`STBlob`|
+|`TxnSignature`| |`string`|`STBlob`|
+|`Signers`| |`array`|`STArray`|
+
+Either the `SigningPubKey` and `TxnSignature` fields must be included, or the `Signers` field.
+
+#### 2.5.1. `Account`
+
+This is an account that has at least one inner transaction.
+
+#### 2.5.2. `SigningPubKey` and `TxnSignature`
+
+These fields are included if the account is signing with a single signature (as opposed to multi-sign). They sign the `Flags` and `TxnIDs` fields.
+
+#### 2.5.3. `Signers`
+
+This field is included if the account is signing with multi-sign (as opposed to a single signature). It operates equivalently to the [`Signers` field](https://xrpl.org/docs/references/protocol/transactions/common-fields/#signers-field) used in standard transaction multi-sign. This field holds the signatures for the `Flags` and `TxnIDs` fields.
+
+### 2.6. Metadata
+
+The inner transactions will be committed separately to the ledger and will therefore have separate metadata. This is to ensure better backwards compatibility for legacy systems, so they can support `Batch` transactions without needing any changes to their systems.
+
+For example, a ledger that only has one `Batch` transaction containing 2 inner transactions would look like this:
+```
+[
+  OuterTransaction,
+  InnerTransaction1,
+  InnerTransaction2
+]
+```
+
+#### 2.6.1. Outer Transactions
+
+Each outer transaction will only contain the metadata for its sequence and fee processing, not for the inner transaction processing.
+
+There will also be a list of which transactions were actually processed, which is useful for the `ONLYONE` and `UNTILFAILURE` modes, since those may only process a subset of transactions, and for debugging with all modes. This section will be called `BatchExecutions`.
+
+It will contain a list of objects that have the following fields for every transaction that is processed (successfully or unsuccessfully):
+
+|FieldName | Required? | JSON Type | Internal Type |
+|:---------|:-----------|:---------------|:------------|
+|`TransactionHash`|✔️|`string`|`STUInt256`|
+|`InnerResult`|✔️|`string`|`STBlob`|
+
+Some important things to note:
+* It is possible that all transactions will not be included in this list. For example, when using the `ONLYONE` mode, if the first transaction succeeds, then the rest of the transactions will not even be processed. 
+* Transactions will only be included in the ledger if their result code is `tesSUCCESS` _and_ if the outer transaction has a result code of `tesSUCCESS`. For example, if the `ALLORNOTHING` mode is used and _any_ inner transaction fails, _none_ of the inner transactions will be included in the ledger.
```

Comment:
This aspect of `Batch` transactions appears to diverge from the current understanding of the XRPL blockchain. If a non-batch transaction fails with a `tec...` error, such a transaction is included in the ledger. [Docs](https://xrpl.org/docs/references/protocol/transactions/transaction-results/tec-codes#tec-codes)

However, suppose an Outer-Batch transaction comprises of 10 inner transactions and the `tfAllOrNothing` mode. If it fails on the 9th inner transaction, then none of the transactions are included on the ledger. Am I understanding this correctly?

Reply:
If the `Batch` transactions `tec...` codes are not included in the ledger, it leads to a break in the provenance of blockchain transactions. A few drops of XRP is irrevocably lost as transaction fees and we can never trace the reasons for such decrease in an account's balance.

Is there any design reason for excluding `tec...` transactions? What are the disadvantages?

Additionally, would the `account_tx` also not display such err-ing Batch transactions? `account_tx` uses a different database for indexing and storing the transactions and I haven't looked at the implementation.

Reply:
>  A few drops of XRP is irrevocably lost as transaction fees

Inner Batch transactions don't have a fee. The only fee comes from the outer transaction, which will still be validated. So no break in the provenance.

> Is there any design reason for excluding tec... transactions? What are the disadvantages?

If, as in your example, the 9th of 10 transactions with `tfAllOrNothing` fails, why is only the 9th transaction validated and not the previous 8? That's not consistent.

> Additionally, would the account_tx also not display such err-ing Batch transactions? account_tx uses a different database for indexing and storing the transactions and I haven't looked at the implementation.

It will show the whole `Batch` transaction as one, the inner transactions will not be displayed differently.

Reply:
> Inner Batch transactions don't have a fee. The only fee comes from the outer transaction, which will still be validated. So no break in the provenance.

I missed this aspect of the spec, this clarifies my doubts 👍 

> 
> > Is there any design reason for excluding tec... transactions? What are the disadvantages?
> 
> If, as in your example, the 9th of 10 transactions with `tfAllOrNothing` fails, why is only the 9th transaction validated and not the previous 8? That's not consistent.

No, I would have preferred to specify the execution status of all the inner transactions (not just a particular failing transaction) in the ledger. This provides the best transparency for debugging purposes.

To explain my stand further, could we execute all the transactions of an `tfAllOrNothing` Batch transaction and specify the transaction-execution-result for each of them? 
> 
> > Additionally, would the account_tx also not display such err-ing Batch transactions? account_tx uses a different database for indexing and storing the transactions and I haven't looked at the implementation.
> 
> It will show the whole `Batch` transaction as one, the inner transactions will not be displayed differently.

No, you are speaking from the perspective of the account which submits the Outer-`Batch` transaction. If I query `account_tx` with one of the inner-`Batch`-transaction account's, what is the expected outcome?

I can wait for the implementation and try-it-out myself, this question is not urgent.

Reply:
> No, I would have preferred to specify the execution status of all the inner transactions (not just a particular failing transaction) in the ledger. This provides the best transparency for debugging purposes.
> To explain my stand further, could we execute all the transactions of an tfAllOrNothing Batch transaction and specify the transaction-execution-result for each of them?

See the section on `BatchExecutions` (in 2.6.1).

> No, you are speaking from the perspective of the account which submits the Outer-Batch transaction. If I query account_tx with one of the inner-Batch-transaction account's, what is the expected outcome?

The inner transactions that are validated will all show up in `account_tx`. Not sure if the failed ones will show up - I don't know how the tracking of "is an account involved in this transaction" works.

> I can wait for the implementation and try-it-out myself, this question is not urgent.

https://dev.to/dangell7/batch-xls-56-is-now-available-for-testing-and-development-18bk